src/ray/util/process.cc

```diff
@@ -405,12 +398,30 @@ void Process::Kill() {
       HANDLE handle = fd != -1 ? reinterpret_cast<HANDLE>(fd) : NULL;
       if (!::TerminateProcess(handle, ERROR_PROCESS_ABORTED)) {
         error = std::error_code(GetLastError(), std::system_category());
+        if (error.value() == ERROR_ACCESS_DENIED) {
+          // This can occur under some circumstances if the process is already terminating
+          DWORD exit_code;
+          if (GetExitCodeProcess(handle, &exit_code) && exit_code != STILL_ACTIVE) {
+            // The process is already terminating, so consider the killing successful
+            error = std::error_code();
+          }
+        }
       }
 #else
-      (void)fd;
-      if (kill(pid, SIGKILL) != 0) {
+      pollfd pfd = {static_cast<int>(fd), POLLHUP};
+      if (fd != -1 && poll(&pfd, 1, 0) == 1 && (pfd.revents & POLLHUP)) {
+        // The process has already died; don't attempt to kill its PID again.
+      } else if (kill(pid, SIGKILL) != 0) {
         error = std::error_code(errno, std::system_category());
       }
+      if (error.value() == ESRCH) {
```

Comment:
Should we be logging a warning in this case?

Reply:
It's treated like any other error, so it'll get logged as an error a little further down. Or did you mean changing the error to a warning in this case? I don't know which one is better because it could be closer to an error than a warning depending on the circumstances under which `Kill()` is called. From a semantic perspective it's erroneous (has a race condition) for a non-owning process to kill another process, but maybe that's unrealistic to consider an error? Let me know if I should demote it to a warning. (Though I'm hoping this case is still unlikely due to the immediate `poll` prior to `kill`.)

Reply:
I see, without reading closely enough I thought this was just getting ignored. I think leaving it as an error is ok.